File: 2022/day06_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    return 14 + next(
        index
        for index, values in enumerate(
            zip(
                content,
                content[1:],
                content[2:],
                content[3:],
                content[4:],
                content[5:],
                content[6:],
                content[7:],
                content[8:],
                content[9:],
                content[10:],
                content[11:],
                content[12:],
                content[13:],
            )
        )
        if not repeating(values)
    )


def repeating(values: tuple[str, ...]) -> bool:
    a, *rest = values
    if not rest:
        return False
    return a in rest or repeating(tuple(rest))


def main() -> None:
    with open("day06.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2022/day06-2.ts
function solution(content: string) {
  return (
    14 +
    content
      .split("")
      .map(
        (_, index, array) =>
          !repeating([
            array[index],
            array[index + 1],
            array[index + 2],
            array[index + 3],
            array[index + 4],
            array[index + 5],
            array[index + 6],
            array[index + 7],
            array[index + 8],
            array[index + 9],
            array[index + 10],
            array[index + 11],
            array[index + 12],
            array[index + 13],
          ])
      )
      .findIndex((v) => v)
  );
}

function repeating(values: string[]): boolean {
  const [first, ...rest] = values;
  if (rest.length === 0) return false;
  return rest.includes(first) || repeating(rest);
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
